File: algorithms/0287.find_the_duplicate_number/find_the_duplicate_number.rs
impl Solution {
    pub fn find_duplicate(nums: Vec<i32>) -> i32 {
        // let sum = (0..nums.len() as i32).sum::<i32>();
        // nums.iter().sum::<i32>() - sum
        let mut number = vec![0; nums.len()];
        for &n in nums.iter() {
            if number[n as usize - 1] == -1 {
                return n;
            } else {
                number[n as usize - 1] = -1
            }
        }
        1
    }
}

File: algorithms/0287.find_the_duplicate_number/find_the_duplicate_number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int fast = 0;
        int slow = 0;
        do {
            fast = nums[nums[fast]];
            slow = nums[slow];
        } while (slow != fast);
        slow = 0;
        while (slow != fast) {
            fast = nums[fast];
            slow = nums[slow];
        }
        return slow;
    }
};